firmware/library/L1_Drivers/pwm.hpp

```diff
@@ -1,13 +1,169 @@
 #include <cstdint>
+#include <cstdio>
+#include "config.hpp"
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/pin_configure.hpp"
+#include "L2_Utilities/macros.hpp"
 
 class PwmInterface
 {
  public:
     virtual void Initialize(uint32_t frequency_hz = 1'000) = 0;
-    virtual void SetDutyCycle(float duty_cycle) = 0;
+    virtual void SetDutyCycle(float percentage) = 0;
     virtual float GetDutyCycle() = 0;
     virtual void SetFrequency(uint32_t frequency_hz) = 0;
     virtual uint32_t GetFrequency() = 0;
+    virtual void EnablePwm(bool enable = true) = 0;
+    virtual uint32_t GetMatchRegister0() = 0;
+};
+
+class Pwm : public PwmInterface
+{
+ public:
+    static constexpr uint8_t kPwmFunction = 0b001;
+    static constexpr uint8_t kGpioFunction = 0b000;
+    static volatile uint32_t * match_register_table[7];
+    static LPC_PWM_TypeDef * pwm1;
+    static LPC_SC_TypeDef * sc;
+    
+    enum SysClk : uint8_t
+    {
+        kActivatePwm0 = (1 << 5),
+        kActivatePwm1 = (1 << 6)
+    };
+    enum PwmConfigure : uint8_t
+    {
+        kTimerMode = (3 << 0),
+        kCounterEnable = (1 << 0),
+        kResetMr0 = (1 << 1),
+        kPwmEnable = (1 << 3)
+    };
+
+    template <unsigned pwm_channel>
+    static Pwm CreatePwm()
+    {
+        static_assert(pwm_channel <= 6,
+                      "Available PWM channels only go up to pin 6.");
+        return Pwm(pwm_channel);
+    }
+
+    explicit constexpr Pwm(uint8_t channel):
+        pwm_(pwm_pin_),
+        pwm_pin_(2, static_cast<uint8_t>(channel - 1)),
+        channel_(channel)
+    {
+    }
+
+    explicit constexpr Pwm(const PinConfigureInterface & pin):
+        pwm_(pin),
+        pwm_pin_(PinConfigure::CreateInactivePin()),
+        channel_(1)
+    {
+    }
+
+    ~Pwm()
+    {
+        pwm_pin_.SetPinFunction(kGpioFunction);
+    }
+
+    void Initialize(uint32_t frequency_hz = 1'000) override
+    {
+        SJ2_ASSERT_FATAL(1 <= channel_ && channel_ <= 6,
+         "Channel must be between 1 and 6.");
+        // DIRECT MANIUPULATION OF SYSTEM CLOCK REGISTERS
+        // WILL REPLACE WITH FUNCTIONS LATER
+        // Enables PWM1 power/clock control bit
+        sc->PCONP |= SysClk::kActivatePwm1;
+        // TODO(delwin-lei): Remove when sysclock complete.
+        // END OF DIRECT MANIPULATION
+
+        // Resets PWMTC on Match with MR0
+        pwm1->MCR |= PwmConfigure::kResetMr0;
+        pwm1->MR0 = config::kSystemClockRate / frequency_hz;
+
+        // Enables PWM TC and PC for counting and enables PWM mode
+        EnablePwm();
+        pwm1->CTCR &= ~PwmConfigure::kTimerMode;
+
+        // Enables PWM[channel] output
+        pwm1->PCR |= PwmOutputEnable(channel_);
+
+        pwm_pin_.SetPinFunction(kPwmFunction);
+    }
+
+    void SetDutyCycle(float percentage) override
+    {
+        SJ2_ASSERT_FATAL(0.0 <= percentage/100.0 && percentage/100.0 <= 1.0,
+            "Percentage of Duty Cycle provided is out of bounds.");
+        EnablePwm(false);
```

Comment:
Are you sure that you need to disable PWM each time you change the duty cycle?

Reply:
Maybe. When I didn't have the EnablePwm function it would stop working when I changed the duty cycle until I set the frequency. This situation happens when I go to a smaller duty cycle than the previous value. I think I can achieve the same result by just resetting the counter but I won't be able to confirm until I test it on an oscilloscope. EnablePwm and resetting the counter are located in the same register. The bit that disables and enables the pwm automatically resets the counter which was why I chose to go with enabling and disabling the pwm.